### PR TITLE
Add ActiveStorage attachment methods

### LINKED ISSUE #148: Generate methods for ActiveStorage attachments
**Describe the bug:**

SorbetRails doesn't currently generate methods for ActiveStorage attachments. For example, in my Rails app users have avatars and games have covers, but neither of these are typed.

**Steps to reproduce:**

```ruby
class User < ApplicationRecord
  has_one_attached :avatar
end
```

```ruby
class Game < ApplicationRecord
  has_one_attached :cover
end
```

In `hidden.rbi` we have some methods like this:

```ruby
module User::GeneratedAssociationMethods
  def avatar(); end

  def avatar=(attachable); end
end
```

**Expected behavior:**

It should return either [`ActiveStorage::Attached::One`](https://api.rubyonrails.org/classes/ActiveStorage/Attached/One.html) or [`ActiveStorage::Attached::Many`](https://api.rubyonrails.org/classes/ActiveStorage/Attached/Many.html) depending on the definition in the model.

This is what I'd expect to see in `game.rbi`:

```ruby
module Game::GeneratedAttributeMethods
  extend T::Sig
  
  sig { returns(T.nilable(ActiveStorage::Attached::One)) }
  def cover; end

  sig { params(attachable: T.untyped).returns(T.untyped) }
  def cover=(attachable); end

  # or, if we had something defined like `has_many_attached :covers`
  sig { returns(T.nilable(ActiveStorage::Attached::Many)) }
  def covers; end

  sig { params(attachable: T.untyped).returns(T.untyped) }
  def covers=(*attachables); end
end
```

I'm not entirely sure what the setters take or return, unfortunately.

**Versions:**
- Ruby: 2.6.2
- Rails: 6.0.0
- Sorbet: 
- Sorbet-Rails: 0.5.3 (master)

### PR DESCRIPTION
**Pull request by connorshea**
Also made some small formatting changes in a few places.

Fixes #148.

`attachment_reflections` wasn't added until 6.0, with https://github.com/rails/rails/pull/33018

So this doesn't support 5.2, and I'm not sure of a good way to get it to work. We can do something like `User.reflect_on_all_associations.filter { |x| x.name.to_s.end_with?('attachment') }`, but that's _really_ gross so I don't want to.

Are we fine just having this only support 6.0?

### CODE REVIEW

File: spec/generators/rails-template.rb [hunk 1]
```diff
@@ -129,6 +129,11 @@ class Wizard < ApplicationRecord
       has_many :spell_books
 
       scope :recent, -> { where('created_at > ?', 1.month.ago) }
+      
+      if ['5.2', '6.0'].include?(ENV["RAILS_VERSION"])
```
Comment (connorshea) [inline spec/generators/rails-template.rb]: I don't love this solution, but it's the least-worst I could come up with. I tried a few other ways to do this, and they were all pretty hacky and bad.

File: lib/sorbet-rails/model_plugins/active_storage_methods.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  sig { implementation.params(root: Parlour::RbiGenerator::Namespace).void }
+  def generate(root)
+    # Check that ActiveStorage the attachment_reflections method exists
+    # It was added in 6.0, so it isn't available for 5.2.
+    return unless defined?(@model_class.attachment_reflections) && @model_class.attachment_reflections.length > 0
+
+    assoc_module_name = self.model_module_name("GeneratedAssociationMethods")
+    assoc_module_rbi = root.create_module(assoc_module_name)
+
+    attachment_reflections = @model_class.attachment_reflections.transform_values { |attachment| attachment.class }
```
Comment (connorshea) [inline lib/sorbet-rails/model_plugins/active_storage_methods.rb]: I guess I need to add `attachment_reflections` to the methods for ActiveRecord::Base in the `activerecord.rbi`?

**Comment by connorshea** [comment]
Actually, here's a bit less disgusting option that works on Rails 5.2:
```ruby
ActiveStorage::Attachment.distinct.pluck(:record_type, :name)
#=> [["Game", "cover"], ["User", "avatar"]]
```

**Comment by connorshea** [comment]
Oh also I have no idea how to distinguish between one and many associations with that ^

**Comment by connorshea** [comment]
Tbh I think I’d rather just support this for Rails 6, Rails 5.2 will require too much bending over backwards

**Comment by manhhung741** [comment]
I’m fine with supporting only 6.0 for now, as ActiveStorage really just
becomes first class API in 6.0. I can try to see if there is any
alternative to support 5.2 later.

File: lib/sorbet-rails/model_plugins/active_storage_methods.rb (hunk 2 above)
Comment (manhhung741) [reply lib/sorbet-rails/model_plugins/active_storage_methods.rb]: Yeah, if there are sorbet typecheck errors without it

File: lib/sorbet-rails/config.rb [hunk 3]
```diff
@@ -46,8 +46,10 @@ def initialize
         :active_record_assoc,
         :active_record_finder_methods,
         :custom_finder_methods,
-        :enumerable_collections,
+        :enumerable_collections
```
Comment (manhhung741) [inline lib/sorbet-rails/config.rb]: undo

File: lib/sorbet-rails/config.rb [hunk 4]
```diff
@@ -46,8 +46,10 @@ def initialize
...
       ]
+
+      @enabled_model_plugins << :active_storage_methods if defined?(T.unsafe(ActiveStorage))
```
Comment (manhhung741) [inline lib/sorbet-rails/config.rb]: Hmm if we only support rails v6.0, we might need to change this condition.

Comment (connorshea) [reply lib/sorbet-rails/config.rb]: This doesn't really cause any problems since the plugin itself checks whether the given model has any `attachment_reflections`

**Comment by connorshea** [comment]
Hmm, running the update spec script doesn't seem to change anything? Any ideas why the tests would be failing here?

**Comment by manhhung741** [comment]
> Hmm, running the update spec script doesn't seem to change anything

Does the script run into any problem when regenerating data? It seems to remove `module Wizard::GeneratedAssociationMethods` entirely in your diff. I bet it run into a problem and skip the plugin. Try turning `allow_failure` to false in the `model_rbi_formatter`
```
  def run_plugins(plugins, generator, allow_failure: true)
```

**Comment by connorshea** [comment]
Changing `allow_failure` doesn't seem to do anything either 🤔

**Comment by connorshea** [comment]
The diff it expects is weird too:

```diff
@@ -105,28 +105,28 @@
module Wizard::GeneratedAssociationMethods
  extend T::Sig

-  sig { returns(::ActiveStorage::Attachment::ActiveRecord_Associations_CollectionProxy) }
+  sig { returns(ActiveRecord::Associations::CollectionProxy) }
  def hats_attachments; end

-  sig { params(value: T.any(T::Array[::ActiveStorage::Attachment], ::ActiveStorage::Attachment::ActiveRecord_Associations_CollectionProxy)).void }
+  sig { params(value: T.any(T::Array[T.untyped], ActiveRecord::Associations::CollectionProxy)).void }
  def hats_attachments=(value); end

-  sig { returns(::ActiveStorage::Blob::ActiveRecord_Associations_CollectionProxy) }
+  sig { returns(ActiveRecord::Associations::CollectionProxy) }
  def hats_blobs; end

-  sig { params(value: T.any(T::Array[::ActiveStorage::Blob], ::ActiveStorage::Blob::ActiveRecord_Associations_CollectionProxy)).void }
+  sig { params(value: T.any(T::Array[T.untyped], ActiveRecord::Associations::CollectionProxy)).void }
  def hats_blobs=(value); end

-  sig { returns(T.nilable(::ActiveStorage::Attachment)) }
+  sig { returns(T.nilable(T.untyped)) }
  def school_photo_attachment; end

-  sig { params(value: T.nilable(::ActiveStorage::Attachment)).void }
+  sig { params(value: T.nilable(T.untyped)).void }
  def school_photo_attachment=(value); end

-  sig { returns(T.nilable(::ActiveStorage::Blob)) }
+  sig { returns(T.nilable(T.untyped)) }
  def school_photo_blob; end

-  sig { params(value: T.nilable(::ActiveStorage::Blob)).void }
+  sig { params(value: T.nilable(T.untyped)).void }
  def school_photo_blob=(value); end

  sig { returns(::SpellBook::ActiveRecord_Associations_CollectionProxy) }
```

It generates `sig { returns(::ActiveStorage::Attachment::ActiveRecord_Associations_CollectionProxy) }` when I try to update the test data, but the tests seem to generate `sig { returns(ActiveRecord::Associations::CollectionProxy) }`?

**Comment by manhhung741** [comment]
It must be having conflict with the association plugin hmm

**Comment by connorshea** [comment]
Bizarrely, `spec/test_data/v5.2/expected_wizard_wo_spellbook.rbi` gets the correct sigs for `hats_attachment` but `spec/test_data/v5.2/expected_wizard.rbi` does not :?

File: spec/generators/rails-template.rb (hunk 1 above)
Comment (connorshea) [reply spec/generators/rails-template.rb]: I've replaced this with a hack to get it to only add the lines for 5.2 and 6.0, but it's pretty gross. 🤷‍♂

**Comment by connorshea** [comment]
ActiveStorage attachments and blobs are polymorphic, is that why it's acting so weird?

**Comment by connorshea** [comment]
It looks like `relation_should_be_untyped?` is doing different things sometimes. Would `DISABLE_DATABASE_ENVIRONMENT_CHECK=1` matter?

**Comment by connorshea** [comment]
Oh, actually when creating new instances of `SorbetRails::ModelRbiFormatter` the specs print `Note: Unable to create new instance of ActiveStorage::Attachment` locally but they're generated fine in CI. 🤔

**Comment by connorshea** [comment]
@manhhung741 check d567df4ce7159bdebbe9aa0d83ac340084ce4b93 and see how much you hate it as a solution here :P

File: lib/sorbet-rails/model_plugins/active_storage_methods.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  sig { params(model_class: T.class_of(ActiveRecord::Base), available_classes: T::Set[String]).void }
+  def initialize(model_class, available_classes)
+    super
+  end
+
+  sig { implementation.params(root: Parlour::RbiGenerator::Namespace).void }
+  def generate(root)
+    # Check that ActiveStorage the attachment_reflections method exists
+    # It was added in 6.0, so it isn't available for 5.2.
+    return unless defined?(@model_class.attachment_reflections) && @model_class.attachment_reflections.length > 0
```
Comment (manhhung741) [inline lib/sorbet-rails/model_plugins/active_storage_methods.rb]: Interesting that `defined?` work here. I'd think `respond_to?` or `method_defined?` makes more sense

File: spec/generators/rails-template.rb [hunk 6]
```diff
@@ -128,7 +134,7 @@ class Wizard < ApplicationRecord
       has_one :wand
       has_many :spell_books
 
-      scope :recent, -> { where('created_at > ?', 1.month.ago) }
+      scope :recent, -> { where('created_at > ?', 1.month.ago) }#{attachments}
```
Comment (manhhung741) [inline spec/generators/rails-template.rb]: Is this change intentional? Looks like `#{attachments}` should be removed
```suggestion
      scope :recent, -> { where('created_at > ?', 1.month.ago) }
```

**Comment by manhhung741** [comment]
I think I know what's wrong with the spec. This spec is what was failing. It actually uses the same expected file with a full run spec, which has the attachment & blob classes in "available_classes".
```
  it 'generates correct rbi file for Wizard' do
    formatter = SorbetRails::ModelRbiFormatter.new(Wizard, Set.new(['Wizard', 'Wand', 'SpellBook']))
    expect_match_file(
      formatter.generate_rbi,
      'expected_wizard.rbi',
    )
  end
```
I think your fix is correct.

File: spec/generators/rails-template.rb (hunk 6 above)
Comment (connorshea) [reply spec/generators/rails-template.rb]: This is indeed correct, it’s the hack I used to generate the model with active storage methods in 5.2 and 6.0. You can’t put it on a new line or it causes blank lines in older versions

Comment (manhhung741) [reply spec/generators/rails-template.rb]: Can we put it in a new line? I'd prefer having an extra line than having attachments here on this line. People will be confused by it later for sure.

File: lib/sorbet-rails/model_plugins/active_storage_methods.rb (hunk 5 above)
Comment (connorshea) [reply lib/sorbet-rails/model_plugins/active_storage_methods.rb]: Should I change it?

**Comment by connorshea** [comment]
@manhhung741 I've rebased and addressed all your feedback.

**Comment by manhhung741** [comment]
Thanks! I’ll take a look and merge later today
